### PR TITLE
[VLM] Support cos sin cache for Qwen3-VL & GLM-4.1V

### PR DESCRIPTION
**Pull request by yuan-luo**
## Motivation

Support cos sin cache for Qwen3-VL & GLM-4.1V.

This PR refactors the rotary positional embedding (RoPE) implementation to expose an explicit cosine/sine cache interface and reuse it across the 2D vision RoPE code path. Rather than recomputing frequencies and repeatedly calling cos()/sin(), we precompute and cache the 1D cosine/sine tables once, then index into this cache for both text RoPE and the 2D grid RoPE used by the vision encoder.

Step 1 refactored Qwen3-VL and GLM-4.1V.

Before PR: 490us
image

After PR: 186us
image

Inspired by https://github.com/vllm-project/vllm/pull/28798 & https://github.com/vllm-project/vllm/pull/28962.

```
$SGLANG_USE_CUDA_IPC_TRANSPORT=1 SGLANG_VLM_CACHE_SIZE_MB=512 python -m sglang.launch_server --model-path /home/admin/Qwen3-VL-8B-Instruct/ --host 0.0.0.0 --port 30000 --trust-remote-code --tp-size 2 --enable-cache-report --log-level info --max-running-requests 48 --mem-fraction-static 0.7 --chunked-prefill-size 8192  --attention-backend flashinfer --mm-attention-backend fa3 --log-level debug --log-requests --log-requests-level 1
```

```
$bash bench_local_video.sh 
{"id":"06642dd08e3542bdb47dff2ec8609978","object":"chat.completion","created":1765823917,"model":"auto","choices":[{"index":0,"message":{"role":"assistant","content":"视频里的招牌上写着“小鞋匠洗鞋”，并附有服务内容“修复 上色 保养”和手机号码“15295211190”。","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":"stop","matched_stop":151645}],"usage":{"prompt_tokens":12009,"total_tokens":12050,"completion_tokens":41,"prompt_tokens_details":{"cached_tokens":3},"reasoning_tokens":0},"metadata":{"weight_version":"default","e2e_latency":2586.2960815429688,"ttft_latency":2586.3027572631836,"queue_latency":1.3520624488592148}}
real    0m2.593s
user    0m0.001s
sys     0m0.003s
{"id":"e108ca7921db464990ad4cf7365d86cc","object":"chat.completion","created":1765823918,"model":"auto","choices":[{"index":0,"message":{"role":"assistant","content":"视频里的招牌上写着：\n\n**小鞋匠洗鞋**\n\n此外，招牌上还有以下信息：\n\n- **修复 上色 保养**\n- **手机号码：15295211190**\n\n招牌的背景是黄色的，文字是黑色的，顶部有一个白色鞋子的图案。招牌旁边还有一个红色的波浪形遮阳篷。","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":"stop","matched_stop":151645}],"usage":{"prompt_tokens":12009,"total_tokens":12087,"completion_tokens":78,"prompt_tokens_details":{"cached_tokens":12008},"reasoning_tokens":0},"metadata":{"weight_version":"default","e2e_latency":868.2491779327393,"ttft_latency":723.18434715271,"queue_latency":1.3202261179685593}}
real    0m0.875s
user    0m0.000s
sys     0m0.004s
```

## Accuracy Tests

```
$python -m sglang.launch_server --model-path Qwen/Qwen3-VL-8B-Instruct --host 0.0.0.0 --port 30000 --trust-remote-code --tp-size 2 --enable-cache-report --log-level info --max-running-requests 48 --mem-fraction-static 0.7 --chunked-prefill-size 8192 --attention-backend fa3 --mm-attention-backend fa3 
```

```
$python3 -m lmms_eval --model openai_compatible --model_args model_version=Qwen/Qwen3-VL-8B-Instruct   --tasks mmmu_val   --batch_size 16
```
Baseline:
```
openai_compatible (model_version=Qwen/Qwen3-VL-8B-Instruct), gen_kwargs: (), limit: None, num_fewshot: None, batch_size: 16
| Tasks  |Version|Filter|n-shot| Metric |   |Value |   |Stderr|
|--------|------:|------|-----:|--------|---|-----:|---|------|
|mmmu_val|      0|none  |     0|mmmu_acc|↑  |0.5122|±  |   N/A|
```

PR:
```
openai_compatible (model_version=Qwen/Qwen3-VL-8B-Instruct), gen_kwargs: (), limit: None, num_fewshot: None, batch_size: 16
| Tasks  |Version|Filter|n-shot| Metric |   |Value |   |Stderr|
|--------|------:|------|-----:|--------|---|-----:|---|------|
|mmmu_val|      0|none  |     0|mmmu_acc|↑  |0.5156|±  |   N/A|
```

## Benchmarking and Profiling
8xH20
Qwen3-VL-8B-Instruct
TTFT speedup 2%.

Server:
```
$python -m sglang.launch_server \
  --model-path /home/admin/Qwen3-VL-8B-Instruct \
  --host 0.0.0.0 \
  --port 30000 \
  --trust-remote-code \
  --tp-size 1 \
  --enable-cache-report \
  --max-running-requests 128 \
  --mem-fraction-static 0.7 \
  --chunked-prefill-size 8192 \
  --attention-backend fa3 \
  --mm-attention-backend fa3 \
  --log-level debug \
  --log-requests \
  --log-requests-level 1
```

Client:
```
$python3 -m sglang.bench_serving \
  --backend sglang-oai-chat \
  --dataset-name image \
  --num-prompts 256 \
  --apply-chat-template \
  --random-input-len 128 \
  --random-output-len 1 \
  --image-resolution 560x560 \
  --image-format jpeg \
  --image-count 1 \
  --image-content random \
  --random-range-ratio 0.1 \
  --port 30000 \
  --max-concurrency 32
```

```
Baseline
============ Serving Benchmark Result ============
Backend:                                 sglang-oai-chat
Traffic request rate:                    inf       
Max request concurrency:                 32        
Successful requests:                     256       
Benchmark duration (s):                  20.42     
Total input tokens:                      104053    
Total input text tokens:                 20597     
Total input vision tokens:               83456     
Total generated tokens:                  256       
Total generated tokens (retokenized):    256       
Request throughput (req/s):              12.54     
Input token throughput (tok/s):          5094.98   
Output token throughput (tok/s):         12.54     
Peak output token throughput (tok/s):    32.00     
Peak concurrent requests:                64        
Total token throughput (tok/s):          5107.51   
Concurrency:                             31.03     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   2475.23   
Median E2E Latency (ms):                 2485.64   
---------------Time to First Token----------------
Mean TTFT (ms):                          2475.22   
Median TTFT (ms):                        2485.63   
P99 TTFT (ms):                           3237.06   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          0.00      
Median TPOT (ms):                        0.00      
P99 TPOT (ms):                           0.00      
---------------Inter-Token Latency----------------
Mean ITL (ms):                           0.00      
Median ITL (ms):                         0.00      
P95 ITL (ms):                            0.00      
P99 ITL (ms):                            0.00      
Max ITL (ms):                            0.00      
==================================================

PR:
============ Serving Benchmark Result ============
Backend:                                 sglang-oai-chat
Traffic request rate:                    inf       
Max request concurrency:                 32        
Successful requests:                     256       
Benchmark duration (s):                  20.24     
Total input tokens:                      104050    
Total input text tokens:                 20593     
Total input vision tokens:               83457     
Total generated tokens:                  256       
Total generated tokens (retokenized):    256       
Request throughput (req/s):              12.65     
Input token throughput (tok/s):          5139.84   
Output token throughput (tok/s):         12.65     
Peak output token throughput (tok/s):    32.00     
Peak concurrent requests:                64        
Total token throughput (tok/s):          5152.49   
Concurrency:                             30.72     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   2429.14   
Median E2E Latency (ms):                 2453.28   
---------------Time to First Token----------------
Mean TTFT (ms):                          2429.13   
Median TTFT (ms):                        2453.27   
P99 TTFT (ms):                           3259.75   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          0.00      
Median TPOT (ms):                        0.00      
P99 TPOT (ms):                           0.00      
---------------Inter-Token Latency----------------
Mean ITL (ms):                           0.00      
Median ITL (ms):                         0.00      
P95 ITL (ms):                            0.00      
P99 ITL (ms):                            0.00      
Max ITL (ms):                            0.00      
==================================================
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/layers/attention/vision.py [hunk 1]
```diff
@@ -740,10 +742,32 @@ def forward(
...
+            cos = rotary_pos_emb_cos
+            sin = rotary_pos_emb_sin
+
+            if cos.size(-1) * 2 == self.head_size:
+                cos = torch.cat([cos, cos], dim=-1)
+                sin = torch.cat([sin, sin], dim=-1)
+
+            q, k = apply_rotary_pos_emb(q, k, cos, sin)
+            q = q.view(original_shape)
+            k = k.view(original_shape)
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/srt/layers/attention/vision.py]: Refactored.

File: python/sglang/srt/models/qwen3_vl.py [hunk 2]
```diff
@@ -434,6 +452,15 @@ def fast_pos_embed_interpolate(self, grid_thw):
...
         return patch_pos_embeds
 
+    def compute_attn_mask_seqlen(
+        self,
+        cu_seqlens: torch.Tensor,
+    ) -> torch.Tensor:
+        max_seqlen = torch.zeros([], device=cu_seqlens.device)
+        # TODO(yuan-luo): check support flash attn.
+        # max_seqlen = (cu_seqlens[1:] - cu_seqlens[:-1]).max()
+        return max_seqlen
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/srt/models/qwen3_vl.py]: Refactored.

File: python/sglang/srt/models/qwen3_vl.py [hunk 3]
```diff
@@ -442,26 +469,37 @@ def forward(
...
+            grid_thw_list = grid_thw
+            grid_thw = torch.tensor(grid_thw, dtype=torch.int32)
+        else:
+            grid_thw_list = grid_thw.tolist()
+
         pos_embeds = self.fast_pos_embed_interpolate(grid_thw)
         x += pos_embeds
-        rotary_pos_emb = self.rot_pos_emb(grid_thw)
-
         seq_len, _ = x.size()
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/srt/models/qwen3_vl.py]: Removed.

**Review by BBuf — APPROVED** [review]
Good job. LGTM.

**Review by JustinTong0323 — APPROVED** [review]
Verified, LGTM
